mri/scripts/gridsearch.py

```diff
@@ -0,0 +1,188 @@
+# -*- coding: utf-8 -*-
+##########################################################################
+# pySAP - Copyright (C) CEA, 2017 - 2018
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+import itertools
+
+from joblib import Parallel, delayed
+import numpy as np
+
+
+def _reconstruct_case(case, key_names, init_classes,
+                      kspace_data, fourier_kwargs):
+    """Internal Function to carry out reconstruction for a
+    special case. This function pulls in appropriate keyword arguments
+    from input and declares appropriate Linear, Fourier and Regularizer
+    Operators. These operators later are used to create the image model
+    by defining the reconstructor. Then the reconstruction is carried and
+    results are returned
+
+    Parameters
+    ----------
+    case: list of parameters
+        keyword arguments for different operators for reconstruction
+    key_names: list of strings
+        names of keyword arguments specified for reconstruction in 'case'
+    init_classes: list of classes
+        classes for varoius operators. must be in format
+        [linear, regularizer, reconstructor]
```

Comment:
I have been thinking about this, the other alternative is to provide strings that specify if it is cartesian/non-cartesian, what kind of reconstruction etc, and thereby map them, but code maintenance wise its too much of overhead.

Reply:
One advantage is that this init_classes is created by the programmer and not the user, so I think this should be fine, for what its worth, this is an internal function and user cannot access this, do you still find possible errors? 

Another option is to pass all classes individual, which bloats up the number of parameters

Reply:
I think I would be happier to have a higher number of parameters. 
I understand this is a programmer-used function, but still, let's make our lives easier.

If you have a lot of parameters, you can always use `kwargs` when calling the function to make sure not to mix anything up.

It might also be an indicator of bad design. For example, couldn't we imagine instantiating all the objects outside the parallel call and passing them to the job. This might also resolve the whole NFFT not being pickled stuff no?

EDIT
------

Actually sorry got mixed up, the last comment is shit, the NFFT problem was actually the other way around.

Reply:
And the "parameter bloating" is also true for the `grid_search_test_case` param.

Reply:
>And the "parameter bloating" is also true for the grid_search_test_case param.

I dont get you here, so you want to bloat up this too? Like regularizer_params, optimizer_params etc? 
If this is True, maybe we can combine Initclasses and params and pass them in dictionary? (Similar to how user provides it, just that they hold just the parameters and not of list for searching)

Reply:
yes or even better, a test case class holding all of this. This way you don't have string related errors that are not caught by linters.

Reply:
Agreed, let me see how to reformat, this is probably the best way.

Reply:
Done